### PR TITLE
Fix sig suggestions when keyword arguments are present

### LINKED ISSUE #3656: Recent keyword args changes broke sig suggestion
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0Aextend%20T%3A%3ASig%0A%0Asig%20%7Bparams(x%3A%20Integer%2C%20y%3A%20String).void%7D%0Adef%20foo(x%3A%2C%20y%3A)%0Aend%0A%0Adef%20bar(x%2C%20y)%0A%20%20foo(x%3A%20x%2C%20y%3A%20y)%0Aend)

```ruby
# typed: strict
extend T::Sig

sig {params(x: Integer, y: String).void}
def foo(x:, y:)
end

def bar(x, y)
  foo(x: x, y: y)
end
```

#### Observed output

```
editor.rb:8: This function does not have a sig https://srb.help/7017
     8 |def bar(x, y)
        ^^^^^^^^^^^^^
  Autocorrect: Use `-a` to autocorrect
    editor.rb:8: Insert sig {params(x: String, y: T.untyped).void}

     8 |def bar(x, y)
        ^
Errors: 1
```

#### Expected behavior

The suggested sig should be

```ruby
sig {params(x: Integer, y: String).void}
```

- - -

SigSuggestion.cc is doing something dumb—indexing directly into the `params` vector of the method symbol with an index that comes from a send's arguments array:

https://github.com/sorbet/sorbet/blob/7bd53b68bfb8165a5bcd76057522671cbeecfc46/infer/SigSuggestion.cc#L59-L70

This ignores the fact that a send's keyword args effectively are stored like this:

```
foo(:x, x, :y, y)
```

with `Send::numPosArgs` controlling where the positional arguments end and the keyword key/value pairs start.

### PR DESCRIPTION
**Pull request by elliottt**
Fix signature suggestions for methods that have kwargs sends in their body.

### Motivation

Fixes #3656 

### Test plan

See included automated tests.

### CODE REVIEW

File: infer/SigSuggestion.cc [hunk 1]
```diff
@@ -56,16 +56,34 @@ optional<core::AutocorrectSuggestion::Edit> maybeSuggestExtendTSig(core::Context
...
+                to = argInfo.type;
+                break;
+            }
+        }
+        if (to == nullptr) {
+            return nullptr;
+        }
+    } else {
+        to = args[argId].type;
+    }
```
Comment (elliottt) [inline infer/SigSuggestion.cc:81]: We were only considering positional arguments before as sends previously took all the keyword arguments as a single positional hash. With inlined keyword arguments we can make use of the symbols we parse out of the send.

File: infer/SigSuggestion.cc [hunk 2]
```diff
@@ -89,30 +107,52 @@ void extractSendArgumentKnowledge(core::Context ctx, core::LocOffsets bindLoc, c
...
     auto dispatchInfo = snd->recv.type.dispatchCall(ctx, dispatchArgs);
 
-    int i = -1;
-
     // See if we can learn what types should they have
-    for (auto &arg : snd->args) {
-        i++;
+    bool inKwArgs = false;
+    for (int i = 0, argIdx = 0; i < snd->args.size(); i++, argIdx++) {
+        inKwArgs = inKwArgs || (i >= numPosArgs && !hasKwSplat);
```
Comment (elliottt) [inline infer/SigSuggestion.cc:127]: The previous for loop was incrementing `i` for each element of `snd->args`, which would treat the inlined keyword argument pairs the same way as normal positional arguments.

File: test/cli/suggest-sig/suggest-sig.out [hunk 3]
```diff
@@ -306,7 +306,7 @@ suggest-sig.rb:112: This function does not have a `sig` https://srb.help/7017
      112 |  def self.load_account_business_profile(merchant)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   Autocorrect: Done
-    suggest-sig.rb:112: Inserted `sig {params(merchant: T.untyped).returns(Integer)}
+    suggest-sig.rb:112: Inserted `sig {params(merchant: TestCarash::Merchant).returns(Integer)}
```
Comment (elliottt) [inline test/cli/suggest-sig/suggest-sig.out:309]: This is a case where we weren't able to get information about the `merchant` parameter before, as it was used locally with a keyword args send.

File: infer/SigSuggestion.cc [hunk 4]
```diff
@@ -89,30 +112,52 @@ void extractSendArgumentKnowledge(core::Context ctx, core::LocOffsets bindLoc, c
...
+        optional<core::NameRef> keyword;
+        if (inKwArgs) {
+            if (core::isa_type<core::LiteralType>(snd->args[i].type)) {
+                auto lit = core::cast_type_nonnull<core::LiteralType>(snd->args[i].type);
+                if (lit.literalKind == core::LiteralType::LiteralTypeKind::Symbol) {
+                    keyword = lit.asName(ctx);
+                }
+            }
+
+            // increment over the keyword argument
```
Comment (jez) [inline infer/SigSuggestion.cc:144]:
```suggestion
            // increment over the keyword argument symbol
```

File: infer/SigSuggestion.cc [hunk 5]
```diff
@@ -89,30 +112,52 @@ void extractSendArgumentKnowledge(core::Context ctx, core::LocOffsets bindLoc, c
         snd->receiverLoc,
         snd->argLocs,
     };
+
+    auto numPosArgs = snd->numPosArgs;
+    auto numKwArgs = snd->args.size() - numPosArgs;
+    bool hasKwSplat = numKwArgs & 0x1;
```
Comment (jez) [inline infer/SigSuggestion.cc:118]:
```suggestion
    bool hasKwSplat = numKwArgs % 2 == 1;
```

File: infer/SigSuggestion.cc [hunk 6]
```diff
@@ -56,16 +56,39 @@ optional<core::AutocorrectSuggestion::Edit> maybeSuggestExtendTSig(core::Context
...
+            return nullptr;
+        }
+    } else {
+        auto &argInfo = args[argId];
+        if (argInfo.flags.isKeyword) {
+            return nullptr;
+        }
+
+        to = argInfo.type;
+    }
```
Comment (jez) [inline infer/SigSuggestion.cc:86]: Is it worth leaving a comment here somewhere about what might need to be changed to do something smarter for implicit keyword args? (`i.e., to get `T::Hash[Symbol, Integer]`)

**Review by jez — APPROVED** [review]
Thanks!

File: infer/SigSuggestion.cc (hunk 6 above)
Comment (elliottt) [reply infer/SigSuggestion.cc:86]: Yep, definitely. I'll add one.

**Pull request merged**